components/private_channel/rust/ffi/private-channel/repsys-crypto/src/lib.rs

```diff
@@ -0,0 +1,306 @@
+) -> (Vec<Ciphertext>, Vec<CompactProof>) {
+    let mut randomized_vector: Vec<Ciphertext> = Vec::new();
+    let mut proofs_correct_randomization: Vec<CompactProof> = Vec::new();
+
+    for i in 0..encrypted_checks.len() {
+        let randomizer = Scalar::random(&mut OsRng);
+
+        randomized_vector.push(
+            encrypted_checks[i] * randomizer
+        );
+        let mut transcript = Transcript::new(b"CorrectRandomization");
+        let (proof, _) = dleq::prove_compact(
+            &mut transcript,
+            dleq::ProveAssignments {
+                x: &randomizer,
+                A: &randomized_vector[i].points.0,
+                B: &encrypted_checks[i].points.0,
+                H: &randomized_vector[i].points.1,
+                G: &encrypted_checks[i].points.1,
+            },
+        );
+        proofs_correct_randomization.push(proof);
+    }
+    (randomized_vector, proofs_correct_randomization)
+}
+
+/// Verify Randomization proofs
+pub fn verify_randomization_proofs(
+    encrypted_vector: &Vec<Ciphertext>,
+    randomized_vector: &Vec<Ciphertext>,
+    proofs: &Vec<CompactProof>,
+) -> Result<bool, &'static str> {
+    if encrypted_vector.len() != randomized_vector.len() {
+        println!("Error: Size of encrypted vector must equal the size of the randomized vector. Got {} (enc vec) and {} (part decrypted)", encrypted_vector.len(), randomized_vector.len());
+        return Err("Size of encrypted vector must be the same as randomized vector");
+    }
+
+    if encrypted_vector.len() != proofs.len() {
+        println!("Error: Size of encrypted vector must equal the number of proofs. Got {} (enc vec) and {} (proofs)", encrypted_vector.len(), proofs.len());
+        return Err("Size of encrypted hashes slice must be the same as vector checks");
+    }
+
+    let mut result = true;
+    for i in 0..proofs.len() {
+        let mut transcript = Transcript::new(b"CorrectRandomization");
+        result = result & dleq::verify_compact(
+            &proofs[i],
+            &mut transcript,
+            dleq::VerifyAssignments {
+                A: &randomized_vector[i].points.0.compress(),
+                B: &encrypted_vector[i].points.0.compress(),
+                H: &randomized_vector[i].points.1.compress(),
+                G: &encrypted_vector[i].points.1.compress(),
+            },
+        ).is_ok()
+    }
+    Ok(result)
+}
+
+/// Computes partial decryption of ciphertext and returns a set of proofs of
+/// correct decryption.
+/// The partial decryption is due to the shared key.
+pub fn partial_decryption_and_proof(
+    randomized_vector: &Vec<Ciphertext>,
+    sk: &SecretKey,
+) -> (Vec<Ciphertext>, Vec<CompactProof>) {
+    let partial_decryption: Vec<Ciphertext> = randomized_vector
+        .into_iter()
+        .map(|ciphertext| Ciphertext {
+            pk: ciphertext.pk,
+            points: (ciphertext.points.0, sk.decrypt(ciphertext)),
+        })
+        .collect();
```

Comment:
Thanks for the catch. The idea of the output of this interaction between the user and the server, is that the server should only learn if the encrypted text (at the end of the interaction) is a zero or not. To this end, we want an encryption scheme that allows the server to perform computations over the plaintext, and in fact, learn if the encrypted text equals a certain value. 
To avoid leakage of further information, the user proceeds with the randomisation of the plaintext (more details in a comment below) before proceeding with the decryption.
If we use an IND-CCA2 secure scheme, such as the Cramer-Shoup scheme, this would not allow the server to compute the checks over the ciphertext without breaking the decryption process. 

Do you think it makes sense then, to use the ElGamal scheme in this protocol?

Reply:
It does make sense to use ElGamal but perhaps (not sure if this changed since I last looked) it would be a good idea to rotate the server's key, or even make it ephemeral?

Reply:
We can do that, yes. The way we designed the protocol is such that in the first step the user makes a GET request, and receives the server's key, so it would be straightforward to make the server's key ephemeral (we'll have to do that in the server side). However, I fail to see what would be the benefit of this. Is it to further reduce the possibility of the server's key leakage? In that case, do you think ephemeral keys would be required or a slower key rotation would suffice?